File: server/loans/gw-eligibility/src/main/java/ee/rsx/kata/bank/loans/adapter/eligibility/eligibleperiod/FirstEligiblePeriodAdapter.java
package ee.rsx.kata.bank.loans.adapter.eligibility.eligibleperiod;

import ee.rsx.kata.bank.loans.domain.limits.gateway.DetermineEligiblePeriod;
import ee.rsx.kata.bank.loans.domain.segment.CreditSegment;
import jakarta.inject.Named;

import java.util.Optional;

import static java.util.Optional.*;

@Named
class FirstEligiblePeriodAdapter implements DetermineEligiblePeriod {

  @Override
  public Optional<Integer> forLoan(Integer amount, CreditSegment creditSegment) {
    if (creditSegment.isDebt()) {
      return empty();
    }
    return of(calculateFirstMinimumPeriodEligibleFor(amount, creditSegment));
  }

  private Integer calculateFirstMinimumPeriodEligibleFor(Integer amount, CreditSegment creditSegment) {
    double firstPeriod = Math.floor((double) amount / creditSegment.creditModifier());

    return Double.valueOf(firstPeriod).intValue() + 1;
  }
}

File: server/loans/gw-eligibility/src/main/kotlin/ee/rsx/kata/bank/loans/adapter/eligibility/eligibleperiod/FirstEligiblePeriodAdapter.kt
package ee.rsx.kata.bank.loans.adapter.eligibility.eligibleperiod

import ee.rsx.kata.bank.loans.domain.limits.gateway.DetermineEligiblePeriod
import ee.rsx.kata.bank.loans.domain.segment.CreditSegment
import jakarta.inject.Named
import kotlin.math.floor

@Named
internal class FirstEligiblePeriodAdapter : DetermineEligiblePeriod {

  override fun invoke(forAmount: Int, forSegment: CreditSegment): Int? =
    if (forSegment.isDebt)
      null
    else
      calculateFirstMinimumPeriodEligibleFor(forAmount, forSegment)

  private fun calculateFirstMinimumPeriodEligibleFor(amount: Int, creditSegment: CreditSegment): Int {
    val firstPeriod = floor(amount.toDouble() / creditSegment.creditModifier)
    return firstPeriod.toInt() + 1
  }
}
